bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesReferenceProvider.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * #%L
+ * ACS AEM Commons Bundle
+ * %%
+ * Copyright (C) 2020 Adobe
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.wcm.properties.shared.impl;
+
+import com.adobe.acs.commons.wcm.PageRootProvider;
+import com.day.cq.commons.jcr.JcrConstants;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+import com.day.cq.wcm.api.reference.Reference;
+import com.day.cq.wcm.api.reference.ReferenceProvider;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.jcr.resource.api.JcrResourceConstants;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+
+/**
+ * SharedPropertiesReferenceProvider finds the components that have been shared across the site
+ * using the current resource by going over the components that are present in the current resource
+ * page and its template. After identifying the shared components under the current resource page,
+ * this adds the root page to the list of references that needs to be published.
+ */
+@Component(service = ReferenceProvider.class,
+        configurationPolicy = ConfigurationPolicy.REQUIRE)
+public final class SharedComponentPropertiesReferenceProvider implements ReferenceProvider {
+
+    @org.osgi.service.component.annotations.Reference
+    private PageRootProvider pageRootProvider;
+
+    @Override
+    public List<Reference> findReferences(Resource resource) {
+        List<Reference> references = new ArrayList<>();
+
+        if (!resource.getPath().startsWith("/content")) {
+            return references;
+        }
+
+        ResourceResolver resourceResolver = resource.getResourceResolver();
```

Comment:
Im curious if we should be using a service user for this, to ensure proper permissions to read everything that needs to be read.

Reply:
Could probably make the argument that if the user doesnt have permission to read something in here, they couldn't have edited a related shared property, so _maybe_ this is fine?  Generally speaking i would expect a service user for this type of operation rather than depending on the current user's permissions.